python/tank/util/metrics.py

```diff
@@ -0,0 +1,371 @@
+
+    def __init__(self, tk, metrics_queue, log=None):
+        """Initialize the worker thread.
+
+        :params tk: Toolkit api instance.
+        :params metrics_queue: Holds metrics to log in SG.
+        :params logging.Logger log: Logger for debugging purposes.
+
+        """
+
+        super(_MetricsDispatchWorkerThread, self).__init__()
+
+        self._tk = tk
+        self._log = log
+        self._metrics_queue = metrics_queue
+
+        # don't wait for this thread to exit
+        self.daemon = True
+
+    def run(self):
+        """Runs a loop to dispatch metrics as they're added to the queue."""
+
+        # The thread is a daemon. Let it run for the duration of the process
+        while True:
+
+            # get the next available metric and dispatch it
+            try:
+                metrics = self._metrics_queue._get_metrics(
+                    self.DISPATCH_BATCH_SIZE)
+                if metrics:
+                    self._dispatch(metrics)
+            except Exception, e:
+                if self._log:
+                    self._log.error("Error dispatching metrics: %s" % (e,))
+            finally:
+                # wait before dispatching more metrics
+                time.sleep(self.DISPATCH_INTERVAL)
+
+    def _dispatch(self, metrics):
+        """Dispatch the supplied metric to the sg api registration endpoint.
+
+        :param Metric metrics: The Toolkit metric to dispatch.
+
+        """
+
+        # get this thread's sg connection via tk api
+        sg_connection = self._tk.shotgun
+
+        # handle proxy setup by pulling the proxy details from the main
+        # shotgun connection
+        if sg_connection.config.proxy_handler:
+            opener = urllib2.build_opener(sg_connection.config.proxy_handler)
+            urllib2.install_opener(opener)
+
+        # build the full endpoint url with the shotgun site url
+        url = "%s/%s" % (sg_connection.base_url, self.API_ENDPOINT)
+
+        # construct the payload with the auth args and metrics data
+        payload = {
+            "auth_args": {
+                "session_token": sg_connection.get_session_token()
+            },
+            "metrics": [m.data for m in metrics]
+        }
+        payload_json = json.dumps(payload)
+
+        header = {'Content-Type': 'application/json'}
+        request = urllib2.Request(url, payload_json, header)
+        response = urllib2.urlopen(request)
+
+        # execute the log_metrics core hook
+        self._tk.execute_core_hook(
+            platform_constants.TANK_LOG_METRICS_HOOK_NAME,
+            tk=self._tk,
+            metrics=metrics,
+        )
+
+        if self._log:
+            for metric in metrics:
+                self._log.debug("Logged metric: %s" % (metric,))
+
+
+###############################################################################
+# ToolkitMetric classes and subclasses
+
+class ToolkitMetric(object):
```

Comment:
I was thinking it might be useful to having interface on `metrics` rather than just introspecting the dictionaries. This would let us rework the metric internals if we needed to. Especially if we're exposing them in the core hook, might save some headache down the road. If we add standard fields like timestamps too, then we could expose those. 

That said, I don't feel terribly strong about it.